src/extension/baser.ts

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+import { Uri, window, workspace } from 'vscode';
+import '../shared/extension';
+import { Store } from './store';
+
+export class Baser {
+    constructor(
+        private readonly distinctLocalNames: Map<string, string>,
+        private readonly store: Pick<Store, 'distinctArtifactNames'>) {
+    }
+
+    private basesArtifactToLocal = new Map<string, string>([]) // <artifactPath, localPath>
+    private updateBases(artifact: string[], local: string[]) {
+        const i = Array.commonLength(artifact.slice().reverse(), local.slice().reverse());
+        this.basesArtifactToLocal.set(
+            artifact.slice(0, -i).join('/'),
+            local.slice(0, -i).join('/'));
+    }
+
+    private validatedPathsArtifactToLocal = new Map<string, string>()
+    private validatedPathsLocalToArtifact = new Map<string, string>()
+    private updateValidatedPaths(artifact: string, local: string) {
+        this.validatedPathsArtifactToLocal.set(artifact, local);
+        this.validatedPathsLocalToArtifact.set(local, artifact);
+    }
+
+    // Other possibilities:
+    // * 1 local -> 2 artifacts (localPath still possible exact match of artPath)
+    // * Actual 0 artifact match
+    // Future:
+    // 2:2 matching
+    // Notes:
+    // If 2 logs have the same path, then likely the same (unless the path is super short)
+    // If 2 logs don't have the same path, they can still potentially be the same match
+    public translateLocalToArtifact(localPath: string): string { // Future: Ret undefined when certain.
+        // Need to refresh on PathMap update.
+        if (!this.validatedPathsLocalToArtifact.has(localPath)) {
+            const {file} = localPath;
+            if ((	// If no workspace, then the open docs (at this moment) become the workspace.
+            // Over-assuming the localPath.name is distinct. There could be 2+ open docs with the same name.
+                !workspace.workspaceFolders?.length ||
+                    this.distinctLocalNames.has(file)
+            ) &&
+                this.store.distinctArtifactNames.has(file)) {
+
+                const artifactPath = this.store.distinctArtifactNames.get(file)!; // Not undefined due to surrounding if.
+                this.updateValidatedPaths(artifactPath, localPath);
+                this.updateBases(artifactPath.split('/'), localPath.split('/'));
+            }
+        }
+        return this.validatedPathsLocalToArtifact.get(localPath) ?? localPath;
+    }
+
+    // Hacky: We are using `openTextDocument` to test the existence of documents as VS Code does not provide a dedicated existence API.
+    // The similar Node `fs` API does not resolve custom URI schemes in the same way that VS Code does otherwise we would use that.
+    private async pathExists(path: string) {
```

Comment:
>pathExists [](start = 18, length = 10)

Suggestions, after calling URI.parse(path, /* strict */ true), if the scheme is "file", then use fs.exist(Sync) rather than opening a document to save VSCode resources.

Reply:
Calling `openTextDocument` bugs me too and "feels" slow and this idea crossed my mind. However, in practice, I have not seen a performance concern. Are you experiencing any issues here?

Note: 1) This is effectively called on user action and is not ambiently in the background, 2) we cache so it reduces the calls to this, 3) if the file *does not* exist, I would imagine it is not meaningfully greater cost than `fs.exist`, 4) if the file *does* exist, `openTextDocument` will be called on it subsequently anyway (and VS Code seems to only open once).

---
In reply to: [441632961](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r441632961) [](ancestors = 441632961)

Reply:
Found a `workspace.fs.stat` which works slightly better.

---
In reply to: [442416380](https://github.com/microsoft/sarif-vscode-extension/pull/282#discussion_r442416380) [](ancestors = 442416380,441632961)